[DataPipe] Enforcing single valid iterator for IterDataPipes with single DataPipe as output

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70216
* #70373
* #76999
* #75995
* __->__ #70479
* #76776

This PR introduces the single iterator constraint for `IterDataPipe`, focusing on `IterDataPipe` that produces one single `IterDataPipe` as output. This constraint is necessary because having multiple iterators referencing the same `IterDataPipe` can lead to incoherent internal state (e.g. `shuffler` may not produce sensible outputs), such that we cannot save a snapshot of the DataPipe or have determinism mechanisms.

Fixes part of https://github.com/pytorch/data/issues/45

The expected behavior of single iterator per DataPipe is described in details within the linked issue. Please review the changes in `torch/utils/data/datapipes/_typing.py` to see if the implementation is sensible. The new test `TestIterDataPipeSingletonConstraint` in `test_datapipe.py` illustrates the behaviors that the constraint should impose. Please comment there if any behavior seems unexpected/unclear.

## BC-breaking Note:
* By imposing a single iterator constraint for `IterDataPipe`, each `IterDataPipe` can only have one active iterator at a time. The most recently created iterator is considered to be the active one; all previous ones are invalid. Attempting to use an invalid DataPipe will result in an error. The code example below illustrates this behavior:

**1.11.0 or before**:
```python
source_dp = IterableWrapper(range(10))
it1 = iter(source_dp)
list(it1)  # [0, 1, ..., 9]
it1 = iter(source_dp)
next(it1)  # 0
it2 = iter(source_dp)  
next(it2)  # returns 0
next(it1)  # returns 1
```
**This PR**:
```python
source_dp = IterableWrapper(range(10))
it1 = iter(source_dp)
list(it1)  # [0, 1, ..., 9]
it1 = iter(source_dp)  # This doesn't raise any warning or error
next(it1)  # 0, works because it is a new iterator
it2 = iter(source_dp)
next(it2) # returns 0, invalidates `it1`
next(it1)  # This raises an error
```
* A consequence of this change is that multiple downstream references to the same `IterDataPipe` is not allowed.

**1.11.0 or before**:
```python
source_dp = IterableWrapper(range(10))
zip_dp = source_dp.zip(source_dp)
list(zip_dp)  # [(0, 0), ..., (9, 9)]
```
**This PR**:
```python
source_dp = IterableWrapper(range(10))
zip_dp = source_dp.zip(source_dp)
list(zip_dp)  # This raises an error because there are multiple references to `source_dp`
```
* The recommended workaround is to use `.fork()`. 
```python
source_dp = IterableWrapper(range(10))
dp1, dp2 = source_dp.fork(2)
zip_dp = dp1.zip(dp2)
list(zip_dp)  # [(0, 0), ..., (9, 9)]
```
* Note that: this does **_NOT_** impact `MapDataPipe`, `Dataset`, or `IterableDataset`.

Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

NivekT committed:
[DataPipe] Enforcing single iterator per IterableWrapperIterDataPipe

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33344609).

NivekT committed:
Update on "[DataPipe] Enforcing single iterator per IterableWrapperIterDataPipe"


Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single iterator per IterableWrapperIterDataPipe"


Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single iterator per IterableWrapperIterDataPipe"


Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single iterator per IterableWrapperIterDataPipe"


Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single iterator per IterableWrapperIterDataPipe"


Fixes part of https://github.com/pytorch/data/issues/45

The expected behavior of single iterator per DataPipe is described in that linked issue. Please review the changes in `torch/utils/data/datapipes/_typing.py` to see if the implementation is sensible.

Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes with single DataPipe as output"


Fixes part of https://github.com/pytorch/data/issues/45

The expected behavior of single iterator per DataPipe is described in that linked issue. Please review the changes in `torch/utils/data/datapipes/_typing.py` to see if the implementation is sensible.

Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes with single DataPipe as output"


Fixes part of https://github.com/pytorch/data/issues/45

The expected behavior of single iterator per DataPipe is described in that linked issue. Please review the changes in `torch/utils/data/datapipes/_typing.py` to see if the implementation is sensible.

Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

[ghstack-poisoned]

NivekT commented:
@VitalyFedyunin @ejguan This PR and the next one in the stack are ready for you to have a look. 

I expect some CI breakage since we probably re-used some DataPipes somewhere in our tests. I am in the process of going through and fixing those. I am also going through the domain use cases to see if anything breaks.

ejguan reviewed (commented):
Can I know we need incremental `id` for each `DataPipe`? IIUC, we only need a switch between iterator created and iterator ended.

NivekT commented:
> Can I know we need incremental `id` for each `DataPipe`? IIUC, we only need a switch between iterator created and iterator ended.

In this implementation, the DataPipes don't hold references to active iterators, but only the ID (`int`) of the active iterator. The iterators hold references to the DataPipe.

Whenever `__next__` is called by an iterator, it first checks whether if it is still the active iterator of the DataPipe by comparing the iterator's ID with the DataPipe's ID. The incrementation of the DataPipe's ID invalidates all previous iterators except for the most recently created one.

ejguan commented:
> > Can I know we need incremental `id` for each `DataPipe`? IIUC, we only need a switch between iterator created and iterator ended.
> 
> In this implementation, the DataPipes don't hold references to active iterators, but only the ID (`int`) of the active iterator. The iterators hold references to the DataPipe.
> 
> Whenever `__next__` is called by an iterator, it first checks whether if it is still the active iterator of the DataPipe by comparing the iterator's ID with the DataPipe's ID. The incrementation of the DataPipe's ID invalidates all previous iterators except for the most recently created one.

Do you mean we can still create multiple iterators from a single DataPipe? But, if we start to iterate over one of them, we can not iterate the other ones.
What would be the benefit over directly disallowing creation of multiple iterators from `IterDataPipe`?

ejguan commented:
I see. What would be the behavior of the following code?
```py
dp = IterableWrapper(range(10))
it1 = iter(dp)
next(it1)
it2 = iter(dp)
```
I believe `it2` will invalidate `it1`. But, we need to consider how this is going to affect checkpoint implementation. In order to retrieve the state of DataPipe, the internal state of iterator (like buffer) needs to be attached to datapipe instance. It means we still need to `reset` these internal state when a new iterator is created.

> 1. it may need some extra public API to reset the DataPipe to read from the beginning again (rather than just calling `__iter__` again)

Then, this is going to be a similar problem for both cases. 
- Allowing multiple iterators would require a private API like `_reset_iterator` to clean up these internal state whenever new iterator is created. Otherwise, checkpointing would be an unknown state.
- If we disallow multiple iterators, we have to let users explicitly invoke `reset_iterator` when users want a new iterator.

> 2. it may be annoying notebook mode, such as re-running the following code snippet in a cell:

This is indeed a problem if we disallow multiple iterators.

The main problem in my mind is still the multi-reference scenario. What would be the output if we allow multiple iterators and invalidate the previous iterator?

```
dp = IterableWrapper(range(10))
dp1 = dp.zip(dp)
list(dp1)  # Expected empty list?
```
What do we expect in this case?

NivekT commented:
> I see. What would be the behavior of the following code?
> 
> ```python
> dp = IterableWrapper(range(10))
> it1 = iter(dp)
> next(it1)
> it2 = iter(dp)
> ```
`next(it1)` will return `0`, and `it1` will be invalid once `it2` is created (even if `it2` is not ran).

> I believe `it2` will invalidate `it1`. But, we need to consider how this is going to affect checkpoint implementation. In order to retrieve the state of DataPipe, the internal state of iterator (like buffer) needs to be attached to datapipe instance. It means we still need to `reset` these internal state when a new iterator is created.
> 
> > 1. it may need some extra public API to reset the DataPipe to read from the beginning again (rather than just calling `__iter__` again)
> 
> Then, this is going to be a similar problem for both cases.
> 
> * Allowing multiple iterators would require a private API like `_reset_iterator` to clean up these internal state whenever new iterator is created. Otherwise, checkpointing would be an unknown state.
> * If we disallow multiple iterators, we have to let users explicitly invoke `reset_iterator` when users want a new iterator.

Turns out a lot of our states (with the exception of `demux`, `fork` that use `ChildDataPipe`) are attached to `__iter__`, such as:
https://github.com/pytorch/pytorch/blob/master/torch/utils/data/datapipes/iter/combinatorics.py#L120

I think we only need to implement `reset` for `ChildDataPipe`.

> > 2. it may be annoying notebook mode, such as re-running the following code snippet in a cell:
> 
> This is indeed a problem if we disallow multiple iterators.
> 
> The main problem in my mind is still the multi-reference scenario. What would be the output if we allow multiple iterators and invalidate the previous iterator?
> 
> ```
> dp = IterableWrapper(range(10))
> dp1 = dp.zip(dp)
> list(dp1)  # Expected empty list?
> ```
> 
> What do we expect in this case?

It raises an exception:
```
RuntimeError: This iterator has been invalidated because another iterator has been createdfrom the same IterDataPipe: IterableWrapperIterDataPipe(deepcopy=True, iterable=range(0, 10))
This may be caused multiple references to the same IterDataPipe. We recommend using `.fork()` if that situation arises.
For feedback regarding this single iterator per IterDataPipe constraint, feel free to comment on this issue: https://github.com/pytorch/data/issues/45.
This exception is thrown by __iter__ of IterableWrapperIterDataPipe(deepcopy=True, iterable=range(0, 10))
```

ejguan commented:
> Turns out a lot of our states (with the exception of `demux`, `fork` that use `ChildDataPipe`) are attached to `__iter__`, such as: https://github.com/pytorch/pytorch/blob/master/torch/utils/data/datapipes/iter/combinatorics.py#L120

That's exact problem for checkpoint. Considering the DataPipe is in the middle of iteration and users invoke the `checkpoint` or `state_dict`, how could we retrieve the `buffer` state? It was not possible in current design because `buffer` is attached to iterator rather than `DataPipe` instance. I want to make sure each DataPipe owning a singleton iterator because we can assign this `buffer` to `DataPipe` instance now. We won't be worried about multiple `buffer`s existing for a single `DataPipe` instance.

So, we have to expose `buffer` object to `self`. Then, if we want to reset iterator, we have to reset `buffer`.

Please correct me if I am wrong.

NivekT commented:
Yea I think you are right that all buffers should be tied to the DataPipe (`self`). In that case, all buffer DataPipes will need some steps to reset the buffer whenever the DataPipe is reset.

I think comes down to how the usage feels from the user perspective and what is the API that we will expose to them. I will  put together a design and schedule a time offline to go over this.

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes with single DataPipe as output"


Fixes part of https://github.com/pytorch/data/issues/45

The expected behavior of single iterator per DataPipe is described in that linked issue. Please review the changes in `torch/utils/data/datapipes/_typing.py` to see if the implementation is sensible.

Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes with single DataPipe as output"


Fixes part of https://github.com/pytorch/data/issues/45

The expected behavior of single iterator per DataPipe is described in that linked issue. Please review the changes in `torch/utils/data/datapipes/_typing.py` to see if the implementation is sensible.

Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes with single DataPipe as output"


Fixes part of https://github.com/pytorch/data/issues/45

The expected behavior of single iterator per DataPipe is described in that linked issue. Please review the changes in `torch/utils/data/datapipes/_typing.py` to see if the implementation is sensible.

Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33344609).

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes with single DataPipe as output"


Fixes part of https://github.com/pytorch/data/issues/45

The expected behavior of single iterator per DataPipe is described in that linked issue. Please review the changes in `torch/utils/data/datapipes/_typing.py` to see if the implementation is sensible.

Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes with single DataPipe as output"


Fixes part of https://github.com/pytorch/data/issues/45

The expected behavior of single iterator per DataPipe is described in that linked issue. Please review the changes in `torch/utils/data/datapipes/_typing.py` to see if the implementation is sensible.

Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33344609).

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes with single DataPipe as output"


Fixes part of https://github.com/pytorch/data/issues/45

The expected behavior of single iterator per DataPipe is described in that linked issue. Please review the changes in `torch/utils/data/datapipes/_typing.py` to see if the implementation is sensible.

Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

[ghstack-poisoned]

NivekT commented on test/test_profiler.py:
```diff
@@ -167,9 +167,12 @@ def __iter__(self):
-        it_dp2 = iter(dp2)
-        self.assertEqual(5, it_dp2.get_value(5))
-        self.assertEqual(list(range(10)), list(it_dp2))
+        it_dp2 = iter(dp2)  # This creates a 2nd iterator, hence invalidating `it_dp1` and `dp1`
+        self.assertEqual(5, it_dp2.get_value(5))  # type: ignore[attr-defined]
+        with self.assertRaisesRegex(RuntimeError, "A separate iterator has been created"):
+            next(it_dp2)
+        with self.assertRaisesRegex(RuntimeError, "A separate iterator has been created"):
+            self.assertEqual(list(range(10)), list(it_dp2))
```
The necessary change to this test case here is worth having a closer look.

The basic issue was that within our constraint logic, we decided to increment the counter even when `iter` returns `self`, and this invalidates the source datapipe `dp1` such that calling `next` will trigger an exception

NivekT commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -373,15 +373,53 @@ def _generate_input_args_string(obj):
+    Given an instance of a DataPipe and an iterator ID, check if the IDs match, and if not, raises an exception.
+    """
+    if hasattr(datapipe, "_is_child_datapipe") and datapipe._is_child_datapipe is True:
+        pass  # TODO: Add logic for multiple ChildDataPipe
```
This is added in the next PR within this stack

NivekT commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -422,25 +465,43 @@ def wrap_generator(*args, **kwargs):
-                return IteratorDecorator(iter_ret)
 
-            namespace['__iter__'] = wrap_iter
+            # Note that if the `__next__` and `__iter__` do something completely unrelated? It may cause issue but
+            # the user will be violating the iterator protocol
```
If the `__next__` and `__iter__` do something completely unrelated, it would violate the iterator protocol and the behavior here maybe unexpected. But as long as user sticks with iterator protocol or returns an iterator, everything should be fine.

NivekT commented on test/test_datapipe.py:
```diff
@@ -2155,5 +2161,78 @@ def test_old_dataloader(self):
+            def __next__(self):
+                return next(self.iterable)
+
+        # Functional Test: Check that every `__iter__` call returns the same object
+        source_dp = _CustomIterDP_Self(range(10))
+        it1 = iter(source_dp)
+        self.assertEqual(0, next(it1))
+        self.assertEqual(1, next(source_dp))
+        it2 = iter(source_dp)
```
As discussed offline, we are applying the single iterator constraint even when `iter` returns `self`. This means the users can at most create one iterator from the object.

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33344609).

ejguan commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -390,24 +428,29 @@ def __next__(self):
         @functools.wraps(func)
         def wrap_generator(*args, **kwargs):
             gen = func(*args, **kwargs)
+            datapipe = args[0]
+            iterator_id = _set_datapipe_valid_iterator_id(datapipe)
```
Could we make sure `reset` will be invoked everytime `iter` is called?

NivekT replied:
One idea is to check if the DataPipe `hasattr(dp, "reset")` and call it here if it does. I am open to do that instead of relying on `__iter__` to call `reset` (which is what a PR later in the stack is doing).

ejguan replied:
I prefer to add a noop `reset` to base class and always call it during `__iter__` to save users' effort. And, adding a contribution doc to indicate `reset` is required.

NivekT replied:
Okay, let me try and add that to the PR later in the stack.

ejguan commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -422,25 +465,43 @@ def wrap_generator(*args, **kwargs):
         # IterDataPipe is an iterator with both ``__iter__`` and ``__next__``
-        if '__next__' in namespace:
+        # And ``__iter__`` may or may not return `self`
+        if '__next__' in namespace:  # If `__next__` exists, put a wrapper around it
```
Is there way that we can check if the `__iter__` is returning `self`?
And, it seems we never set `datapipe._valid_iterator_id` to 1 for this case (`_set_datapipe_valid_iterator_id` is never invoked)

NivekT replied:
> Is there way that we can check if the `__iter__` is returning `self`?

I can have another look. A bit tricky since we are within the hook, we might have to call it and examine the output.

> And, it seems we never set `datapipe._valid_iterator_id` to 1 for this case (`_set_datapipe_valid_iterator_id` is never invoked)

Yes, it is only set later by `wrap_iter` if `iter` is called. In theory the user can have two active references (i.e. `dp` and `it = iter(dp)`. This is true unless we ban them from using `iter` completely.

ejguan replied:
> A bit tricky since we are within the hook, we might have to call it and examine the output.

If that's the case, we are probably not going to do so

> Yes, it is only set later by `wrap_iter` if `iter` is called. In theory the user can have two active references (i.e. `dp` and `it = iter(dp)`. This is true unless we ban them from using `iter` completely.

What does that mean? Can you provide an example of expected behavior for such scenario?

NivekT replied:
This test case right here:
https://github.com/pytorch/pytorch/blob/87bba0f26e4463fb766bf736ddc09757b1053370/test/test_datapipe.py#L2188-L2214

Notice that both `source_dp` and `it1` are valid here:
```python
self.assertEqual(0, next(it1))
self.assertEqual(1, next(source_dp))
```

NivekT replied:
Yea, I think when `__iter__` returns `self`, any attempt to update/restrict ID doesn't work too well in notebook mode.

I originally proposed that if `__iter__` returns `self`, we do not need to track any ID, and should just let the users call `__iter__` as many time as they would like. That has a different downside too, since users might accidentally have multiple references to it without realizing.

ejguan replied:
I believe we are proposing a `reset` function, right? Then, whenever `__iter__` is invoked, the `reset` will handle re-initialization for all variables like `buffer`. As long as users have proper `reset` function, it should be fine.

NivekT replied:
The only issue is when they re-run a cell 2 without re-running cell 1 in the example below. An exception will be raised the second time they run it because `dp` will become invalidated. The workaround is to override `reset` such that `_valid_iterator_id` is set back to `None` during reset.

Cell 1:
```python
dp = CustomDP()  # Its `__iter__` returns `self`
```
Cell 2:
```python
it = iter(dp)
```

VitalyFedyunin replied:
Wait, I remember we discussed an alternative to hook __iter__ method and to check if it returns self. And in such a scenario wrap the return into a new IterableContainer of some sort to track validation. Let me know if you need sudocode for an explanation.

NivekT replied:
It does get wrapped around by `IteratorDecorator` which delegates method calls to the iterator object (if `self` is returned, then the DataPipe is the iterator object).

Please look at this comment: https://github.com/pytorch/pytorch/pull/70479#discussion_r874217619

ejguan commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -422,25 +465,43 @@ def wrap_generator(*args, **kwargs):
+                    datapipe = args[0]
+                    # The `_valid_iterator_id` should either be never set (`None`), or set by at most one
+                    # iterator (`0`). Otherwise, it means there are multiple iterators.
+                    if datapipe._valid_iterator_id is not None and datapipe._valid_iterator_id != 0:
+                        raise RuntimeError(
+                            "A separate iterator has been created for this IterDataPipe, but `next()` has been called"
+                            f"on the original IterDataPipe as well: {_generate_iterdatapipe_msg(datapipe)}\n"
+                            "Multiple iterator references to the same IterDataPipe is not allowed. We recommend "
+                            "using `.fork()` if that is necessary.\n"
+                            "For feedback regarding this single iterator per IterDataPipe constraint, feel free "
+                            "to comment on this issue: https://github.com/pytorch/data/issues/45.")
```
Could we combine this part to `_check_iterator_valid`?

NivekT replied:
Done.

ejguan commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -422,25 +465,43 @@ def wrap_generator(*args, **kwargs):
+        def wrap_iter(*args, **kwargs):
+            iter_ret = func(*args, **kwargs)
+            datapipe = args[0]
+            nonlocal iterator_id
```
Noob question: why do we need `nonlocal` here?

NivekT replied:
Just to make it explicit. Probably matters more for the IDE when renaming a variable name.

ejguan replied:
I am still not clear with it. Why is `iterator_id` local variable for the other places?

NivekT replied:
`iterator_id` is a variable tied to the iterator (inside `hook_iterator` and initialized to `None`). 

https://github.com/pytorch/pytorch/blob/87bba0f26e4463fb766bf736ddc09757b1053370/torch/utils/data/datapipes/_typing.py#L438

Since it is used by both the wrapper of `next` and `iter`, it has to be at that location. It is different from `datapipe._valid_iterator_id` which stores the latest, active iterator ID of a given DataPipe.

ejguan replied:
But, both `next` and `iter` function have access to `self` aka `datapipe`.
`__iter__(self)` and `__next__(self)`

ejguan reviewed (commented):
Since this PR introduces BC breaking changes, could you please add a detailed behavior changes in the PR summary? When we do release note, it will save more time.

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes with single DataPipe as output"


This PR introduces the single iterator constraint for `IterDataPipe`, focusing on `IterDataPipe` that produces one single `IterDataPipe` as output. This constraint is necessary because having multiple iterators referencing the same `IterDataPipe` can lead to incoherent internal state (e.g. `shuffler` may not produce sensible outputs), such that we cannot save a snapshot of the DataPipe or have determinism mechanisms.

Fixes part of https://github.com/pytorch/data/issues/45

The expected behavior of single iterator per DataPipe is described in details within the linked issue. Please review the changes in `torch/utils/data/datapipes/_typing.py` to see if the implementation is sensible. The new test `TestIterDataPipeSingletonConstraint` in `test_datapipe.py` illustrates the behaviors that the constraint should impose. Please comment there if any behavior seems unexpected/unclear.

Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

[ghstack-poisoned]

ejguan commented on test/test_datapipe.py:
```diff
@@ -2268,5 +2274,78 @@ def test_old_dataloader(self):
+        self.assertEqual(1, next(it2))
+        with self.assertRaisesRegex(RuntimeError, "This iterator has been invalidated"):
+            next(it1)
+
```
Could you please add a test for a chain of DataPipes?
And, it would be great if this chain of DataPipes contains these different DataPipes.

NivekT replied:
Done.

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -2268,5 +2274,78 @@ def test_old_dataloader(self):
+        with self.assertRaisesRegex(RuntimeError, "This iterator has been invalidated"):
+            next(it1)
+
+        r"""
```
def test_foo?

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -2268,5 +2274,78 @@ def test_old_dataloader(self):
+            def __iter__(self):
+                return self
+
+            def __next__(self):
```
Also need buggy case when it returns a new object, but also have __next__

NivekT replied:
I am adding a buggy case at the bottom. Let me know if it matches what you have in mind.

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -2268,5 +2274,78 @@ def test_old_dataloader(self):
+        it2 = iter(source_dp)
+        with self.assertRaisesRegex(RuntimeError, "This iterator has been invalidated"):
+            next(it1)
+        with self.assertRaisesRegex(RuntimeError, "This iterator has been invalidated"):
```
wait, why it is invalid?

VitalyFedyunin replied:
do we have at least one valid iterator after this?

NivekT replied:
In my original proposal, the users can are allowed to call `__iter__` as many times as they want. But after our offline discussion, we decided to apply the constraint on `__iter__` even when it may be returning `self`. This is because, having multiple active iterators referencing the same underlying object can lead to confusion and bugs. Plus the use case isn't really necessary when you can just reference `dp`. The code snippet below illustrates what the behavior would've been like if it is allowed:
```python
dp = CustomDP(range(10))
next(dp)  # returns 0
it1 = iter(dp)  # returns `self`
next(it1)  # returns 1
it2 = iter(dp)  # returns `self`
next(it2)  # returns 2
next(it1)  # returns 3
next(dp)  # returns 4
```

In the current implementation, calling `next(dp)` raises an error if there is more than 1 iterator created from that DataPipe. This is because the ID that tracks whether the iterator is valid (i.e. `dp._valid_iterator_id`) is tied to the DataPipe (and because it returns `self`, the ID is tied to the iterator as well). I think it has to be the case, otherwise when `next(dp)` is called. It has no idea to what ID to check against.

```python
source_dp = _CustomIterDP_Self(range(10))
it1 = iter(source_dp)
self.assertEqual(0, next(it1))
self.assertEqual(1, next(source_dp))
it2 = iter(source_dp)
with self.assertRaisesRegex(RuntimeError, "This iterator has been invalidated"):
    next(it1)
with self.assertRaisesRegex(RuntimeError, "This iterator has been invalidated"):
    next(it2)
with self.assertRaisesRegex(RuntimeError, "This iterator has been invalidated"):
    next(source_dp)
```

The simplest workaround for users is to re-define `dp` via `dp = CustomDP()` or avoid creating more than one iterator from `dp` by keep using `dp`  as the variable rather than calling `iter(dp)`.

ejguan replied:
Here are the proposal you have
- For `DataPipe` returning `self` in `__iter__`: We disallow users to call `iter` multiple times.
- For `DataPipe` using a generator function as `__iter__`: We are making sure a new iterator is created and previous iterator is invalidated whenever `iter` is called.
As we are making iterator of `DataPiep` singleton (the same behavior), we are introducing a new difference on iterator. I think this is not preferable.

I think as we have `reset` function to help `DataPipe` to reset iterator. Why don't we rely on `reset` to clean up all buffer, etc. for `DataPipe` returning `self` in `__iter__`. And, the `IteratorDecorator` and `datapipe` itself should be able to track the iterator_id to invalidate previous iterator.

NivekT replied:
I can see how that can be confusing.

I think this proposal will be best, given a DataPipe `dp` with `__next__` with iterators `it1` and `it2`:
1. Allows unlimited creation of iterators, but only the latest one is valid/active
2. Doesn't place restriction on `next(dp)` (but check for invalidation when `next(it1)` is called
    - We can't really place restriction on `next(dp)`, because any call to `next(it2)` gets delegated to `next(dp)` 
       within the `IteratorDecorator`

```python
source_dp = _CustomIterDP_Self(range(10))
it1 = iter(source_dp)
self.assertEqual(0, next(it1))
self.assertEqual(1, next(source_dp))
# Only invalidates `it1`, and not `source_dp`. Since methods of `it2` depends on `source_dp` remaining valid
it2 = iter(source_dp)
with self.assertRaisesRegex(RuntimeError, "This iterator has been invalidated"):
    next(it1)
self.assertEqual(0, next(it2))
self.assertEqual(1, next(source_dp))
```


Please examine this test for the full details:
https://github.com/pytorch/pytorch/blob/c77df3ecf5110560e602ff5c4b3b5825ffb168c0/test/test_datapipe.py#L2337-L2390

NivekT commented:
> I suddenly realize there may potentially be another problem. When we invalidate the previous iterator, we don't destroy them. It means the iterator object is never released from memory except the case the `DataPipe` itself is an iterator.

The iterator object should get garbage collected if there is no longer any references to it. Correct?
```python
dp = IterableWrapper(range(10))
it1 = iter(dp)
it2 = iter(dp)  # invalidates `it1`
`it1` can now be gc'ed because there isn't any reference to it
```

ejguan commented:
> The iterator object should get garbage collected if there is no longer any references to it. Correct?
> 
> ```python
> dp = IterableWrapper(range(10))
> it1 = iter(dp)
> it2 = iter(dp)  # invalidates `it1`
> `it1` can now be gc'ed because there isn't any reference to it
> ```

`it1` is a reference to the iterator. Until users explicitly run `del it1` or the process ends, the iterator object won't be gc'ed. It makes no sense for us to keep such reference when we invalidate it.

If possible, we could use the `wrapper` class. Whenever the iterator is invalidated, we should remove the internal `self.iterator` to release the memory.

NivekT commented:
@VitalyFedyunin @ejguan I have addressed the comments and added more test cases.

Let's discuss these two related points if they remain to be unclear/problematic:

https://github.com/pytorch/pytorch/pull/70479#discussion_r872850049
https://github.com/pytorch/pytorch/pull/70479#discussion_r871864125

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes with single DataPipe as output"


This PR introduces the single iterator constraint for `IterDataPipe`, focusing on `IterDataPipe` that produces one single `IterDataPipe` as output. This constraint is necessary because having multiple iterators referencing the same `IterDataPipe` can lead to incoherent internal state (e.g. `shuffler` may not produce sensible outputs), such that we cannot save a snapshot of the DataPipe or have determinism mechanisms.

Fixes part of https://github.com/pytorch/data/issues/45

The expected behavior of single iterator per DataPipe is described in details within the linked issue. Please review the changes in `torch/utils/data/datapipes/_typing.py` to see if the implementation is sensible. The new test `TestIterDataPipeSingletonConstraint` in `test_datapipe.py` illustrates the behaviors that the constraint should impose. Please comment there if any behavior seems unexpected/unclear.

Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes with single DataPipe as output"


This PR introduces the single iterator constraint for `IterDataPipe`, focusing on `IterDataPipe` that produces one single `IterDataPipe` as output. This constraint is necessary because having multiple iterators referencing the same `IterDataPipe` can lead to incoherent internal state (e.g. `shuffler` may not produce sensible outputs), such that we cannot save a snapshot of the DataPipe or have determinism mechanisms.

Fixes part of https://github.com/pytorch/data/issues/45

The expected behavior of single iterator per DataPipe is described in details within the linked issue. Please review the changes in `torch/utils/data/datapipes/_typing.py` to see if the implementation is sensible. The new test `TestIterDataPipeSingletonConstraint` in `test_datapipe.py` illustrates the behaviors that the constraint should impose. Please comment there if any behavior seems unexpected/unclear.

Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

[ghstack-poisoned]

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -2290,5 +2296,162 @@ def test_old_dataloader(self):
+                self.iterable = iter(self.source)
+
+        # Functional Test: Check that every `__iter__` call returns the same object
+        source_dp = _CustomIterDP_Self(range(10))
```
Amazing work! The only thing missing by my opinion is moving all tests commonalities into something like:
`def _test_datapipe_iterators_restrictions(self, dp_graph)`

NivekT replied:
I have refactored the test.

ejguan reviewed (approved):
The behavior based on the test cases look good to me. Could you please also update the docstring for `IterDataPipe` to indicate this singleton behavior?

And, could you please import this PR to Phabricator? We don't need to land from internal but it's good to validate this won't break any internal logic

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes with single DataPipe as output"


This PR introduces the single iterator constraint for `IterDataPipe`, focusing on `IterDataPipe` that produces one single `IterDataPipe` as output. This constraint is necessary because having multiple iterators referencing the same `IterDataPipe` can lead to incoherent internal state (e.g. `shuffler` may not produce sensible outputs), such that we cannot save a snapshot of the DataPipe or have determinism mechanisms.

Fixes part of https://github.com/pytorch/data/issues/45

The expected behavior of single iterator per DataPipe is described in details within the linked issue. Please review the changes in `torch/utils/data/datapipes/_typing.py` to see if the implementation is sensible. The new test `TestIterDataPipeSingletonConstraint` in `test_datapipe.py` illustrates the behaviors that the constraint should impose. Please comment there if any behavior seems unexpected/unclear.

Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

[ghstack-poisoned]

NivekT commented on test/test_datapipe.py:
```diff
@@ -2290,5 +2296,162 @@ def test_old_dataloader(self):
+                self.iterable = iter(iterable)
+
+            def __iter__(self):
+                self.reset()
```
Yea, I had to add that line since the code change related to "automatic reset by `__iter__`" is implemented in a later PR in the stack.

NivekT commented:
> The behavior based on the test cases look good to me. Could you please also update the docstring for `IterDataPipe` to indicate this singleton behavior?
> 
> And, could you please import this PR to Phabricator? We don't need to land from internal but it's good to validate this won't break any internal logic

I can definitely update the docstring. I tried importing yesterday but it didn't work but I can try again.

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes with single DataPipe as output"


This PR introduces the single iterator constraint for `IterDataPipe`, focusing on `IterDataPipe` that produces one single `IterDataPipe` as output. This constraint is necessary because having multiple iterators referencing the same `IterDataPipe` can lead to incoherent internal state (e.g. `shuffler` may not produce sensible outputs), such that we cannot save a snapshot of the DataPipe or have determinism mechanisms.

Fixes part of https://github.com/pytorch/data/issues/45

The expected behavior of single iterator per DataPipe is described in details within the linked issue. Please review the changes in `torch/utils/data/datapipes/_typing.py` to see if the implementation is sensible. The new test `TestIterDataPipeSingletonConstraint` in `test_datapipe.py` illustrates the behaviors that the constraint should impose. Please comment there if any behavior seems unexpected/unclear.

Differential Revision: [D33344609](https://our.internmc.facebook.com/intern/diff/D33344609)

[ghstack-poisoned]

b0noI commented:
@pytorchbot revert this, broke internal tests, escalated to Kevin Tse [ktse@fb.com]

ejguan commented:
@b0noI If this PR is reverted, we might need to revert the whole stack.

b0noI commented:
landing it internally and we will do fix forward

NivekT commented:
Thank you!

ejguan commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -373,15 +373,58 @@ def _generate_input_args_string(obj):
+    r"""
+    Given an instance of a DataPipe and an iterator ID, check if the IDs match, and if not, raises an exception.
+    """
+    msg = ("This iterator has been invalidated because another iterator has been created"
+           f"from the same IterDataPipe: {_generate_iterdatapipe_msg(datapipe)}\n"
```
nit: There is a missing space between `created` and `from` in the Error message. cc: @NivekT
